### PR TITLE
Add a validation error if memory is too low.

### PR DESCRIPTION
**Pull request by brendandburns**
Noticed this while implementing the memory resources e2e test.  Right now, if memory is < 4Gb we accept, schedule and then fail the pod.

### CODE REVIEW

**Comment by brendandburns** [comment]
@dchen1107

**Comment by dchen1107** [comment]
yeah, docker failed creation on limit < 4M here :-). But should you update the API with this?

**Comment by brendandburns** [comment]
@bgrant0607 @kprobst 
What do you folks think from a usability perspective.  The error is kind of annoying, but @dchen1107 is correct, it is Docker specific.

I'm ok with either approach.

**Comment by bgrant0607** [comment]
In general, there's quantization of memory due to page and/or NUMA-node sizes. We've experienced quite a bit of pain with different quantization in the scheduler than on the node internally.

Requiring 4MiB increments wouldn't be unreasonable, but would be kind of annoying for kick-the-tires users. It would probably be more understandable for users that are carefully managing use of node resources, but I don't think our users are interested in that degree of resource management yet.

Given that we don't support multiple QoS levels yet and don't reserve resources for every last byte (e.g., the OS), I'd be inclined to just round up the enforced limit in Kubelet.

**Comment by brendandburns** [comment]
@bgrant0607 Docker currently requires 4MiB increments, but we don't fail validation up front, we only do it via Pending and an event.

This PR adds an up-front validation failure.

**Comment by dchen1107** [comment]
I don't think docker requires 4MiB increments today yet. 4MiB is minimal memory allocation for a docker container. I am ok with the changes here, all I were asking is just change API to reflect this change.

File: pkg/api/validation/validation.go [hunk 1]
```diff
@@ -1432,6 +1432,10 @@ func ValidateResourceRequirements(requirements *api.ResourceRequirements) errs.V
 		}
 		allErrs = append(allErrs, errs...)
 	}
+	quantity, found := requirements.Limits[api.ResourceMemory]
+	if found && quantity.Value() < 4*1024*1024 {
+		allErrs = append(allErrs, errs.NewFieldInvalid("resources.limits[memory]", quantity.String(), "Memory must be at least 4Mb"))
```
Comment (bgrant0607) [inline pkg/api/validation/validation.go]: MiB

**Comment by erictune** [comment]
Is this targeting v1?

File: pkg/api/validation/validation.go (hunk 1 above)
Comment (brendandburns) [inline pkg/api/validation/validation.go]: done.

**Comment by brendandburns** [comment]
that is @bgrant0607 's call.

**Comment by bgrant0607** [comment]
This can wait.

**Comment by dchen1107** [comment]
@brendandburns are you going to continue processing this pr, or close it?

**Comment by bgrant0607** [comment]
@vishh is this a libcontainer limitation, a kernel limitation, or what? And is it a minimum value, or minimum granularity?

**Comment by vishh** [comment]
The 4MB minimum limit is in place because the docker binary is essentially a busybox and, during container init time docker binary is invoked in init mode and it requires up to 4MB. If and when docker moves to a nsinit exec model, this limit can be lowered further. 
Will setting up a default limit range help here in addition to the validation logic?

**Comment by vishh** [comment]
This PR LGTM.

**Comment by bgrant0607** [comment]
Ok.

Please add a comment with Vish's explanation. It will otherwise be hard for someone to figure out why this is here later, after various refactorings, code moves, etc.

**Comment by vishh** [comment]
Ping @brendandburns

**Comment by vishh** [comment]
Ping.

**Comment by brendandburns** [comment]
@vishh do you have a PR that superseeds this one?  Or shall I bring this one back from the dead.

Let me know

**Comment by vishh** [comment]
Not yet. If this can be rebased easily go for it. If not, I can create a
new PR.

**Comment by brendandburns** [comment]
@vishh rebased (finally) please take another look.

**Comment by vishh** [comment]
LGTM. Thanks @brendandburns!

**Comment by jlowdermilk** [comment]
@k8s-bot test this

**Comment by jlowdermilk** [comment]
This pr fails to build: https://travis-ci.org/kubernetes/kubernetes/builds/95021317

**Comment by wojtek-t** [comment]
@brendandburns - can you please rebase? This doesn't compile.

**Comment by wojtek-t** [comment]
@brendandburns - can you please rebase? This doesn't compile.

**Comment by wojtek-t** [comment]
@brendandburns - can you please rebase?

**Comment by gmarek** [comment]
@k8s-bot test this

**Comment by gmarek** [comment]
`_output/local/go/src/k8s.io/kubernetes/pkg/api/validation/validation.go:1799: undefined: errs`

**Comment by eparis** [comment]
@k8s-bot test this issue: #IGNORE

Tests have been pending for 24 hours

**Comment by wojtek-t** [comment]
@brendandburns - can you please rebase?

**Comment by wojtek-t** [comment]
And fix test problems?

**Comment by wojtek-t** [comment]
Since @brendandburns didn't fix the problem for more than a month, I'm closing this PR.

**Pull request closed without merging**